### ISSUE TITLE
`fix_noise_scheduler_betas_for_zero_terminal_snr` should come before `prepare_scheduler_for_custom_training`

### ISSUE THREAD

**Issue by jihnenglin** [kohya-ss/sd-scripts#1263]
I think the correct implementation should be like this.
```python
noise_scheduler = DDPMScheduler(
    beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear", num_train_timesteps=1000, clip_sample=False
)
if args.zero_terminal_snr:
    custom_train_functions.fix_noise_scheduler_betas_for_zero_terminal_snr(noise_scheduler)
prepare_scheduler_for_custom_training(noise_scheduler, accelerator.device)
```
The current implementation calculates `noise_scheduler.all_snr` before even fixing the betas in the scheduler. This would make `apply_snr_weight` use the wrong SNR when training zero terminal SNR models with min SNR gamma.

**Comment by jihnenglin** [comment]
Also, this change would cause `apply_snr_weight` (without v-prediction) and `apply_debiased_estimation` to break with a divide by zero error when `zero_terminal_snr` is True. I've just recently started experimenting with zero terminal SNR models, so I'm not entirely sure how to address that issue. I noticed that @drhead seems to have quite a bit of experience in training these types of models. Would you mind sharing your thoughts or any advice you might have on this topic?

**Comment by drhead** [comment]
The difference won't be very significant either way.  It's technically incorrect, but I wouldn't worry about it too much.

At least based on some of my ongoing experiments with adaptive loss weights, the bigger problem is that those ends go towards zero.  Learned timestep weighting schemes will arrive at something like this (ignore the resolution bit and also the waviness of the line): 
![loss_weights_14099_297d6ed00bea96ef0f01](https://github.com/kohya-ss/sd-scripts/assets/1313496/dc95900e-c84e-4312-a2a8-726e1f90f845)
which suggests that the ideal timestep weighting is one that does not weigh any timesteps as orders of magnitude less important than others (😱).  Which does make much more intuitive sense, but before I suggest changes I would want to do a symbolic regression first.  In the mean time, I would say that the answer is that neither is really correct.

**Comment by jihnenglin** [comment]
Thank you! I was getting underwhelming results from my experiments on zero terminal SNR models, and I figured there might be something wrong with the loss scaling. 😅 By the way, how did you get this graph? Is this graph describing a zero terminal SNR model with v loss? I'd be very appreciative if you could link me any sources that I can learn about this adaptive loss weighting learning process.

**Comment by jihnenglin** [comment]
This is very interesting! Thank you again for sharing your amazing findings with me. I'm looking forward to your PR.

**Comment by jihnenglin** [comment]
![line-graph](https://github.com/kohya-ss/sd-scripts/assets/9023825/61141600-1f67-46c4-b175-da717053f865)
I'd like to provide an update on this. The original schedule, represented by the red line, used a minimum signal-to-noise ratio (SNR) gamma of 5, v-prediction, and a zero terminal SNR. The blue line is a new schedule inspired by the learned timestep weighting suggested by @drhead. The results are promising after around 4,500 steps. Notice that the scales at both ends are no longer close to zero, which should help the model learn effectively across all timesteps. I believe this is the right direction, though further testing may be needed.

**Comment by jihnenglin** [comment]
@drhead Thank you! I'll definitely try it later. And sorry for the confusion, I think it have converged way earlier, the 4.5 k step is just the first checkpoint I saved and tested.
Is the learning rate 0.005 good even for fine-tuning the entire model? I'm scared.

**Comment by drhead** [comment]
>  Is the learning rate 0.005 good even for fine-tuning the entire model? I'm scared.

Oh, no, no, you want that just for the loss weight MLP itself.  It's using EDM2-style weight normalization, so it can handle very high learning rates and a lack of weight decay because its activations can't grow almost unbounded like they do on most models.  Use your normal learning rate for the actual diffusion model/LoRA, and set this model's parameters to use the higher learning rate and zero weight decay.

**Comment by jihnenglin** [comment]
Oh, thanks for clearing it up for me! These are super helpful.

**Comment by jihnenglin** [comment]
@drhead This may be an irrelevant issue but most samplers give me weird results at lower (20-30) sampling steps for my v-prediction model. And they all work perfectly fine at higher (> 50) sampling steps! DDIM and Euler are the only ones that give decent results with lower steps. Have you ever had this problem before?

**Comment by drhead** [comment]
> @drhead This may be an irrelevant issue but most samplers give me weird results at lower (20-30) sampling steps for my v-prediction model. And they all work perfectly fine at higher (> 50) sampling steps! DDIM and Euler are the only ones that give decent results with lower steps. Have you ever had this problem before?

Very common issue.  One thing I recommend for mitigating it is setting sigma_max to a lower value like 160 (zsnr options on A1111 and ComfyUI default to ~4500, Diffusers defaults to 4096), though this will not use the terminal timestep and is not ideal.  I recently submitted a PR to A1111 with a scheduler that I find is better from a recent paper: https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/15608
